Add documentation for moderated sessions (#9425)
gravitational/teleport: docs/pages/access-controls/guides/moderated-sessions.mdx

---
title: Moderated Sessions
description: Moderated Sessions
h1: Moderated Sessions
---

## Introduction

> Moderated sessions allows Teleport administrators to

Reviewer: ```suggestion
Moderated Sessions allows Teleport administrators to
```

define requirements for other users to be present in a Server or Kubernetes Access session. Depending on the requirements, these users can observe the session in real time, participate in the session, and terminate the session at will.

### Use cases

Moderated Sessions are useful in the following scenarios:

> - When you have stringent security requirements and need to have people watching over user-initiated sessions to a set of servers.

Reviewer: ```suggestion
- When you have stringent security requirements and need to have people watching over user-initiated sessions on a set of servers.
```

- When you want to share a terminal with someone else to be able to instruct or collaborate.

## Policies

Moderated Sessions makes use of RBAC policies to allow for fine grained control over
who can join a session and who is required to be present to start one.

The system is based around **require policies** and **allow policies**. 

Require policies define a set of conditions that must be a met for a session to start or run.
A minimum of one policy from each relevant role the user has must match for the session to start.

Allow policies are used to define what sessions a user can join
and under what conditions they may join a session.

## Configuring Moderated Sessions

### `require_session_join`

> 

Reviewer: ```suggestion

`require_session_join` configures the require policies for a role. The value of 
`require_session_join` is a list of objects with the following fields.
```
Reading this now, I think we need some kind of introduction for this section.

#### Options

The following are required options for `require_session_join`:

|Option|Type|Description|
|---|---|---|
|`name`|String|The name of the require policy|
|`filter`|[Filter](./moderated-sessions.mdx#filters)|An expression that, if it evaluates to true for a given user, enables the user to be present in a Moderated Session|
|`kinds`|`[]`[Session kind](./moderated-sessions.mdx/#session-kinds)|The kind of session that the policy applies to|
|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that applies to the user joining the Moderated Session under this policy|

> |`count`|[Participant mode](./moderated-sessions.mdx#require-policy-count)|The amount of users that need to match the filter expression to satisfy the policy|

Reviewer: ```suggestion
|`count`|Integer|The number of users that need to match the filter expression to satisfy the policy|
```

#### Example

> The policy below specifies that a given role

Reviewer: ```suggestion
The policy below specifies that the `prod-access` role
```

must have a minimum of two users with the role `auditor` and the mode `moderator` present in the session
to start it. The policy applies to SSH and Kubernetes sessions only.

When a user with this require policy starts a session, it will be pending
until the policy is fulfilled. 

```yaml
kind: role
metadata:
  name: prod-access
spec:
  allow:
    require_session_join:
      - name: Auditor oversight
        filter: 'contains(user.roles, "auditor")'
        kinds: ['k8s', 'ssh']
        modes: ['moderator']
        count: 2
```

### `join_sessions`

> 

Reviewer: ```suggestion

`join_sessions` configures the allow policies for a role. The value of 
`join_sessions` is a list of objects with the following fields.
```
Reading this now, I think we need some kind of introduction for this section.

#### Options

The following are required options for `join_sessions`:

|Option|Type|Description|
|---|---|---|
|`name`|String|The name of the allow policy|
|`roles`|[]String|A list of names for Teleport roles that this policy applies to. Users with this role are eligible to join a Moderated Session under this policy.|
|`kinds`|`[]`[Session kind](./moderated-sessions.mdx#session-kinds)|The kind of session that the policy applies to|

> |`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that paplies to the user joining the Moderated Session under this policy|

Author: A suggestion was made by @ptgott here to allow for filter expressions instead of a roles list if I read the comment in the conceptual guide PR correctly. I reasoned this in the original RFD and ended up sticking with the roles list to make the intended way to construct access hierarchies/systems using these policies more explicit. I feel like a filter expression gives flexibility that doesn't give us any better capabilities in return.

#### Example

The following allow policy attaches to the role `auditor` and allows one to join
SSH and Kubernetes sessions started by a user with the role `prod-access` as a moderator or observer.

```yaml
kind: role
metadata:
  name: auditor
spec:
  allow:
    join_sessions:
      - name: Auditor oversight
        roles : ['prod-access']
        kinds: ['k8s', 'ssh']
        modes: ['moderator', 'observer']
```

> ### Filters

Reviewer: Should we introduce all of the functions/operators in filter expressions here?

Author: Added

> A filter expression or predicate is a string statement used to define logic based on a set of input variables.

Reviewer: ```suggestion
A filter expression is a string statement used to define logic based on a set of input variables.
```
We only say "filter expression" later on, so I thought we could remove "predicate" to simplify things.

Filter expressions  allow for more detailed control.
`user.name == "adam" || contains(user.roles, "cs-observe")` is a simple example of a filter expression
that evaluates to true if the user is Adam or if the user has the trait `cs-observe`.

> Require policies can specify which users they consider as valid with a filter expression. They work like

Reviewer: ```suggestion
Require policies can specify which users they consider as valid with a filter expression.
```
Suggested removing this sentence since we don't have good docs elsewhere on filter expressions.

> other filters used within Teleport such with access requests. The filter context has a `user` object defined

Reviewer: ```suggestion
 The filter context has a `user` object defined
```

Suggested removing this sentence since we don't have good docs elsewhere on filter expressions.

which the set fields `roles` and `name`.

### Session kinds

Require and allow policies have to specify which sessions they apply to. Valid options are `ssh` and `k8s`.

- `ssh` policies apply to all SSH sessions towards a node running the Teleport SSH server.

> - `k8s` policies apply to all kubernetes sessions to clusters connected to teleport.

Reviewer: ```suggestion
- `k8s` policies apply to all Kubernetes sessions on clusters connected to Teleport.
```

To work around this, RBAC rules are ignored for users that only have V4 roles (`version: v4` in the role specification).
New roles are created as V5. V4 roles are upgraded when they are modified in the UI.
If a user has any attached V5 roles (`version: v5` in the role specification), the new RBAC access checks will be enforced.

## MFA-based presence

When `per_session_mfa` is set to `true` via [role or cluster settings](../../access-controls/guides/per-session-mfa.mdx), Teleport enforces
MFA-based presence checks for moderators.
This requires that all moderators wishing to join have a configured U2F or WebAuthn MFA token.

Every 30 seconds, Teleport will issue a prompt to the user in the terminal, asking them
to press their MFA token in the next 15 seconds. This will happen continously during the session
and exists so that moderators are always present and watching a given session.

If no MFA input is received within 60 seconds, the user is kicked
from the session which may pause it, if RBAC policies are no longer met.

## Session invites

Moderatated Sessions tracks sessions via `SessionTracker` resource. When starting

> an interactive SSH or Kubernetes session using `tsh ssh` or `tsh kube exec` respectively,

Reviewer: I wasn't sure how the mention of the `SessionTracker` resource relates to the CLI guidance in the next sentence.

The system is based around **require policies** and **allow policies**. 

Require policies define a set of conditions that must be a met for a session to start or run.
A minimum of one policy from each relevant role the user has must match for the session to start.

Allow policies are used to define what sessions a user can join
and under what conditions they may join a session.

## Configuring Moderated Sessions

### `require_session_join`

#### Options

The following are required options for `require_session_join`:

|Option|Type|Description|
|---|---|---|
|`name`|String|The name of the require policy|
|`filter`|[Filter](./moderated-sessions.mdx#filters)|An expression that, if it evaluates to true for a given user, enables the user to be present in a Moderated Session|
|`kinds`|`[]`[Session kind](./moderated-sessions.mdx/#session-kinds)|The kind of session that the policy applies to|
|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that applies to the user joining the Moderated Session under this policy|
|`count`|Integer|The number of users that need to match the filter expression to satisfy the policy|

> Filter expressions allow for more detailed control over the scope of an allow policy or require policy.

Reviewer: On second thought, this and the next paragraph make more sense where they originally were, in the "Filters" section. This paragraph is a good example of using filters—which we need in the "Filters" section—and the paragraph beginning "Require policies can specify" only makes sense if you understand the filter syntax, i.e., at the end of the "Filters" section.

However, when we move these paragraphs  back to the "Filters" section, maybe we could use this order of sentences, since the example filter illustrates the `user.roles` and `user.name` fields (we'll need to unescape the code fences).

```markdown
Filter expressions allow for more detailed control over the scope of an allow policy or require policy.

Require policies can specify which users they consider as valid with a filter expression.
The filter context has a `user` object defined with the set fields `roles` and `name`.

Here is an example of a filter expression that evaluates to true if the user is Adam or if the user has the trait `cs-observe`:

\`\`\`
equals(user.name, "adam") || contains(user.roles, "cs-observe")
\`\`\`

```

The filter context has a `user` object defined with the set fields `roles` and `name`.

#### Example

The policy below specifies that the `prod-access` role
must have a minimum of two users with the role `auditor` and the mode `moderator` present in the session
to start it. The policy applies to SSH and Kubernetes sessions only.

When a user with this require policy starts a session, it will be pending
until the policy is fulfilled. 

```yaml
kind: role
metadata:
  name: prod-access
spec:
  allow:
    require_session_join:
      - name: Auditor oversight
        filter: 'contains(user.roles, "auditor")'
        kinds: ['k8s', 'ssh']
        modes: ['moderator']
        count: 2
```

### `join_sessions`

#### Options

The following are required options for `join_sessions`:

|Option|Type|Description|
|---|---|---|
|`name`|String|The name of the allow policy|
|`roles`|[]String|A list of names for Teleport roles that this policy applies to. Users with this role are eligible to join a Moderated Session under this policy.|
|`kinds`|`[]`[Session kind](./moderated-sessions.mdx#session-kinds)|The kind of session that the policy applies to|

> |`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that paplies to the user joining the Moderated Session under this policy|

Reviewer: ```suggestion
|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that applies to the user joining the Moderated Session under this policy|
```

```yaml
kind: role
metadata:
  name: auditor
spec:
  allow:
    join_sessions:
      - name: Auditor oversight
        roles : ['prod-access']
        kinds: ['k8s', 'ssh']
        modes: ['moderator', 'observer']
```

### Filters

A filter expression is a string statement used to define logic based on a set of input variables.
The filter expressions follow a restricted subset of Go syntax and have the environment supports
the following functions and operators:
- `contains(set, item)`: Returns true if the item is in the set, otherwise false. The set can be a list or an array
- `equals(a, b)`: Returns true if the two values are equal, otherwise returns false.
- `![expr]`: Negates a boolean expression.
- `[expr] && [expr]`: Performs a logical AND on two boolean expressions.
- `[expr] || [expr]`: Performs a logical OR on two boolean expressions.

Filter expressions allow for more detailed control over the scope of an allow policy or require policy.
`equals(user.name, "adam") || contains(user.roles, "cs-observe")` is a simple example of a filter expression
that evaluates to true if the user is Adam or if the user has the trait `cs-observe`.

> Require policies can specify which users they consider as valid with a filter expression.

Reviewer: Should we move this and the next sentence to the relevant table row in the section on configuring require policies?

 The filter context has a `user` object defined

> which the set fields `roles` and `name`.

Reviewer: ```suggestion
with the set fields `roles` and `name`.
```

define requirements for other users to be present in a Server or Kubernetes Access session.

Moderated Sessions are useful in the following scenarios:
- When you have stringent security requirements and need to have people watching over user-initiated sessions to a set of servers.
- When you want to share a terminal with someone else to be able to instruct or collaborate.

## Core Features

With Moderated Sessions you can:
- Join SSH and Kubernetes sessions from the terminal as an observer, moderator or peer.
- Require auditors to be present in certain sessions.
- Allow people to join your session to observe what you are doing in real time.

## Getting Started

In this guide we'll be using Moderated Sessions to define required participants for a Kubernetes session.
This guide assumes an already running Teleport cluster with a connected Kubernetes cluster named `minikube`.

<Admonition type="tip">
  If you don't have a cluster that meets the aforementioned requirements,
  you can create one by following [this guide](https://goteleport.com/docs/kubernetes-access/getting-started/).
</Admonition>

### Step 1/4: Create the prod-access role

> ```code

Reviewer: This snippet is going to look a bit odd until [this issue](https://github.com/gravitational/next/issues/811) is fixed, but I don't think we should block on that issue.

Author: Agreed, not much we can do here except continue
